### ISSUE TITLE
Consider generalizing Lambda Graph's collection types

### ISSUE THREAD

**Issue by joshsh** [CategoricalData/hydra#87]
Hydra's formal data model (Lambda Graph) currently includes type and term constructors for two special kinds of recursive data structures: lists and maps. Since uniqueness of keys is technically not enforced, maps can be considered as syntactic sugar for lists of pairs. Optionals are also supported.

Now that these types are fully supported in both Haskell and Java, consider generalizing them, perhaps even introducing general-purpose support for functors, applicatives, and monads. The current support for lists and maps can serve as a regression test for the more generic solution. Perhaps even Hydra's built-in Flow monad can be brought into a generalized Lambda Graph data model.

**Comment by joshsh** [comment]
Bumping. Still a good thing to consider prior to Hydra 1.0, though only after closing the loop.

**Comment by joshsh** [comment]
Deferred for 0.16, though not closed. Additional term and type constructors can be added non-destructively, and the original constructors gradually deprecated, if we decide to change anything.

**Comment by joshsh** [comment]
Closing with a verdict, following a design investigation against the current implementation (2026; the issue's 2021 premises no longer hold — in particular, `Flow` has since been removed from the kernel in favor of the abstract `effect<>` constructor and `hydra.lib.effects`).

**Verdict: generalize capability, not representation.** The collection types stay as dedicated structural variants of `Type`/`Term`; what changes (in future work) is the type system's *capability* to abstract over them.

1. **Representation.** A census of every special-casing site found the cost modest and concentrated: ~400 LOC of per-collection rules duplicated between inference and checking, ~30 LOC of unification arms, small per-coder encode arms. Against that, the structural form is load-bearing: the adapter framework classifies and rewrites collections per target language *because* "being a list type" is a syntactic fact, and coders map each variant to a native container. Demoting collections to applied nominal constructors was rejected for 1.0. The expressibility gap (`m a` cannot unify with `list<int32>` today) is addressed instead by #226's solver-internal spine normal form: built-ins act as nominal constants during unification only, structural variants stay canonical at rest, zero wire-format change.

2. **Type classes.** Delegated to #581 (instance records). One hard finding from a live spike: **forall-typed record fields do not type-check, by design** — Hydra admits only inferrable types (prenex polymorphism), so instance members must be ordinary top-level polymorphic definitions, with instance records carrying names/metadata rather than rank-2 fields. Also: `maps.map` supports a lawful `Functor (Map k)` instance (the `ordering k` constraint rides the instance scheme), `sets.map` does not (its constraint falls on the ∀-bound codomain variable); an explicit-dictionary `OrdFunctor` is the deferred hatch.

3. **The monad question.** The monadic structure that actually recurs (lists, optionals, eithers, effects) is already served by per-collection primitives; do-notation term forms were considered and **rejected** — the kernel gains no sugar, and any future surface convenience belongs to #497.

4. **Wiki.** The decision is recorded under [Design § Why collection types are built in](https://github.com/CategoricalData/hydra/wiki/Design#why-collection-types-are-built-in) and the inferrability principle under [Concepts § Key design principles](https://github.com/CategoricalData/hydra/wiki/Concepts#key-design-principles).
